File: src/streaming/ndjson.mjs
/**
 * NDJSON (Newline Delimited JSON) utilities
 * https://github.com/ndjson/ndjson-spec
 */

/**
 * Parse a single NDJSON line
 * @param {Object} options - Options
 * @param {string} options.line - Line to parse
 * @returns {Object|null} Parsed object or null if invalid
 */
export function parseNdjsonLine(options) {
  const { line } = options;
  const trimmed = (line || '').trim();

  // Empty lines are ignored in NDJSON
  if (!trimmed) {
    return null;
  }

  // Must start with { or [ to be valid JSON
  if (!trimmed.startsWith('{') && !trimmed.startsWith('[')) {
    return null;
  }

  try {
    return JSON.parse(trimmed);
  } catch {
    return null;
  }
}

/**
 * Stringify an object to NDJSON line
 * @param {Object} options - Options
 * @param {Object} options.value - Value to stringify
 * @param {boolean} [options.compact] - Use compact JSON (no indentation)
 * @returns {string} NDJSON line (with trailing newline)
 */
export function stringifyNdjsonLine(options) {
  const { value, compact = true } = options;

  if (value === null || value === undefined) {
    return '';
  }

  const json = compact
    ? JSON.stringify(value)
    : JSON.stringify(value, null, 2);

  return json + '\n';
}

/**
 * Parse multiple NDJSON lines
 * @param {Object} options - Options
 * @param {string} options.data - Data containing multiple lines
 * @returns {Object[]} Array of parsed objects
 */
export function parseNdjson(options) {
  const { data } = options;
  const lines = (data || '').split('\n');
  const results = [];

  for (const line of lines) {
    const parsed = parseNdjsonLine({ line });
    if (parsed !== null) {
      results.push(parsed);
    }
  }

  return results;
}

/**
 * Stringify multiple objects to NDJSON
 * @param {Object} options - Options
 * @param {Object[]} options.values - Values to stringify
 * @param {boolean} [options.compact] - Use compact JSON
 * @returns {string} NDJSON string
 */
export function stringifyNdjson(options) {
  const { values, compact = true } = options;

  if (!values || !Array.isArray(values)) {
    return '';
  }

  return values
    .map(value => stringifyNdjsonLine({ value, compact }))
    .join('');
}

File: rust/src/streaming/ndjson.rs
//! NDJSON (Newline Delimited JSON) utilities
//! https://github.com/ndjson/ndjson-spec

use serde_json::Value;

/// Parse a single NDJSON line
///
/// # Arguments
/// * `line` - Line to parse
///
/// # Returns
/// Parsed JSON value or None if invalid
pub fn parse_ndjson_line(line: &str) -> Option<Value> {
    let trimmed = line.trim();

    // Empty lines are ignored in NDJSON
    if trimmed.is_empty() {
        return None;
    }

    // Must start with { or [ to be valid JSON
    if !trimmed.starts_with('{') && !trimmed.starts_with('[') {
        return None;
    }

    serde_json::from_str(trimmed).ok()
}

/// Stringify an object to NDJSON line
///
/// # Arguments
/// * `value` - Value to stringify
/// * `compact` - Use compact JSON (no indentation)
///
/// # Returns
/// NDJSON line with trailing newline
pub fn stringify_ndjson_line(value: &Value, compact: bool) -> String {
    if value.is_null() {
        return String::new();
    }

    let json = if compact {
        serde_json::to_string(value).unwrap_or_default()
    } else {
        serde_json::to_string_pretty(value).unwrap_or_default()
    };

    format!("{}\n", json)
}

/// Parse multiple NDJSON lines
///
/// # Arguments
/// * `data` - Data containing multiple lines
///
/// # Returns
/// Vector of parsed JSON values
pub fn parse_ndjson(data: &str) -> Vec<Value> {
    data.lines()
        .filter_map(|line| parse_ndjson_line(line))
        .collect()
}

/// Stringify multiple objects to NDJSON
///
/// # Arguments
/// * `values` - Values to stringify
/// * `compact` - Use compact JSON
///
/// # Returns
/// NDJSON string
pub fn stringify_ndjson(values: &[Value], compact: bool) -> String {
    values
        .iter()
        .map(|v| stringify_ndjson_line(v, compact))
        .collect()
}

// Tests are in rust/tests/streaming_tests.rs
